### DISCUSSION TITLE
import of ForceAtlas2 layout

### DISCUSSION THREAD

**Discussion by CTMMS** [graphology/graphology#523] [Q&A] [answered]
Can someone tell me how i can implement the forceAtlas2 layout from the graphology-library.js release file? i can't seem to figure out how to call any of the functions/constructors declared in it in my project. i keep getting an error telling me that they can't be defined

**Comment by Yomguithereal** [comment]
It's not really possible to help you with only this information. Are you properly loading the `web` version of the library? Or are you running a synchronous layout or working from node.js?

**Reply by CTMMS** [reply]
I included the standalone release files for graphology.js and [graphology-library.js](https://github.com/graphology/graphology/releases/tag/0.25.4) in my project. (After trying to install them via node.js and not managing to get rid of the MIME type errors the import statements kept getting.) The [graphology docs](https://graphology.github.io/) said to implement a graph (from graphology.js) using const graph = new graphology.Graph(); which worked perfectly. I can't, however, find a method to similarly implement the graphAtlas2 layout (or anything else for that matter) defined in graphology-library.js file. Whatever call I try, I keep getting the error that forceAtlas2 isn't defined in graphology-library.js. 
I don't have much experience with how these files declare exports or TypeScript and d.ts files which is why I gave up on the package manager method in the first place. Apologies, if I'm not being very clear, I kept trying different things to somehow get anything to work and ended up terribly confused.

Since the code provided in the graphology docs:

```
const positions = forceAtlas2(graph, {
    iterations: 50,
    settings: {
        gravity: 10
    }
});

// To directly assign the positions to the nodes:
forceAtlas2.assign(graph);
```

didn't work for me, I tried to compare the files to figureout a way forward. My best guess so far is that the call likely starts with "graphologyLibrary" but I had no luck with

```
const positions = graphologyLibrary.layoutForceatlas2(graph, {
    iterations: 50,
    settings: sensibleSettings
});

graphologyLibrary.layoutForceatlas2.assign(graph);
```

**Reply by Yomguithereal** [reply, marked as answer]
It should probably be `graphologyLibrary.layoutForceAtlas2`, not `graphologyLibrary.layoutForceatlas2`, notice the lowercased `a` in your example. In any case, don't hesitate to `console.log(graphologyLibrary)` to explore the object and see how you can navigate it to find what you need.

**Comment by CTMMS** [comment]
Thank you so much! That seems to work but also results in another error. Can you give me an idea why this call:
<img width="713" alt="image" src="https://github.com/user-attachments/assets/113395e4-85f8-49a0-9f8c-e3392923963d">
leads to this error:
<img width="795" alt="image" src="https://github.com/user-attachments/assets/86e85a67-662a-4563-a1bd-b010bef98ec8">

**Reply by Yomguithereal** [reply]
So in your example you are actually running the layout twice. `graphologyLibrary.layoutForceatlas2` runs the layout synchronously and returns an object mapping each node to its coordinates, while `graphologyLibrary.layoutForceatlas2.assign` run the layout synchronously then assigns the coordinates to each node as attributes, rather than returning an object.

But your error comes from the fact you did not pass an option object to the `assign` function specifying the number of iterations to run. So it raised an error.

**Reply by CTMMS** [reply]
Oh wow, there is such a thing as too much research. While trying to fix my initial problem, I had gotten so terribly lost that your last very helpful response went entirely over my head. Thank you for bearing with me while I found my way again. Finally getting it to work is a big relief
